### PR TITLE
Feature/issue 2 historical prices endpoint

### PR DESCRIPTION
**Pull request by johncho108**
The historical prices endpoint is working for the most part (see screenshot below). The only thing I'm stuck on is that the [Iex API historical prices endpoint](https://iexcloud.io/docs/api/#historical-prices) allows for an optional "date" path variable: GET /stock/{symbol}/chart/{range}/**{date}**.  For example, /stock/twtr/chart/5d vs. /stock/twtr/chart/date/**20190220**. I tried to add an optional path variable in the IexClient class as follows:

```
@GetMapping(value = {"/stock/{symbol}/chart/{range}", "/stock/{symbol}/chart/{date}"})
  List<IexHistoricalPrice> getHistoricalPrices(@PathVariable("symbol") String symbol,
      @PathVariable("range") String range, @PathVariable(value = "date", required = false) String date);
```

However, I keep getting this error: **Method getHistoricalPrices can only contain at most 1 value field**. After doing some research, it seems that optional parameters may not be supported by Feign. For example, [this stackoverflow post](https://stackoverflow.com/questions/59413035/feign-client-support-for-optional-request-param). I tried some other workarounds (like method overloading), but I couldn't get it to go. Perhaps I've implemented IexRestController and IexService incorrectly?

For the purposes of this exercise, should I keep trying to figure out how to implement that optional "date" @PathVariable or should I move onto Issue #3? And if so, do you have any tips on including the optional @PathVariable? 

Issue 2 Tests

### CODE REVIEW

File: src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+@Builder
+public class IexHistoricalPrice {
+
+  private BigDecimal close;
+  private BigDecimal high;
+  private BigDecimal low;
+  private BigDecimal open;
+  private String symbol;
+  private long volume;
+  private String date;
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java:18]: Why is date a string?

File: src/main/java/org/galatea/starter/domain/IexSymbol.java [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
   private String symbol;
   private String name;
-  private Date date;
+  private String date;
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/domain/IexSymbol.java:13]: Why did you change this object type to a string?

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java [hunk 3]
```diff
@@ -31,8 +32,10 @@
    * @return a list of all IexStockSymbols.
    */
   @GetMapping(value = "${mvc.iex.getAllSymbolsPath}", produces = {MediaType.APPLICATION_JSON_VALUE})
-  public List<IexSymbol> getAllStockSymbols() {
-    return iexService.getAllSymbols();
+  public List<IexSymbol> getAllStockSymbols(
+      @RequestParam(value = "token", defaultValue = "${mvc.iexToken}") final String token
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: This endpoint should not have a token as an optional parameter. Would we want downstream systems to provide a token?

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java [hunk 4]
```diff
@@ -44,8 +47,30 @@
   @GetMapping(value = "${mvc.iex.getLastTradedPricePath}", produces = {
       MediaType.APPLICATION_JSON_VALUE})
   public List<IexLastTradedPrice> getLastTradedPrice(
-      @RequestParam(value = "symbols") final List<String> symbols) {
-    return iexService.getLastTradedPriceForSymbols(symbols);
+      @RequestParam(value = "token", defaultValue = "${mvc.iexToken}") final String token,
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: Same point here

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java [hunk 5]
```diff
@@ -44,8 +47,30 @@
...
+
+  ) {
+    return iexService.getLastTradedPriceForSymbols(token, symbols);
+  }
+
+  /**
+   * Get historical prices given stock and time range.
+   *
+   * @param symbol stock symbol for historical prices
+   * @param range time range for historical prices (e.g. 3m, 6m, 5y)
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/entrypoint/IexRestController.java:57]: Can you add some documentation about the format for date?

File: src/main/java/org/galatea/starter/service/IexClient.java [hunk 6]
```diff
@@ -30,6 +33,24 @@
...
+   * @param symbol stock symbol for historical prices
+   * @param range time range for historical prices (e.g. 3m, 6m, 5y)
+   * @return a list of historical prices for the symbol/range passed in
+   */
+  @GetMapping(value = "/stock/{symbol}/chart/{range}")
+  List<IexHistoricalPrice> getHistoricalPrices(@RequestParam("token") String token,
+      @PathVariable("symbol") String symbol,
+      @PathVariable("range") String range);
 
+  @GetMapping(value = "/stock/{symbol}/chart/{range}/{date}")
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/service/IexClient.java:50]: Can you add Javadoc for this method?

File: src/main/java/org/galatea/starter/service/IexClient.java [hunk 7]
```diff
@@ -30,6 +33,24 @@
...
+  List<IexHistoricalPrice> getHistoricalPrices(@RequestParam("token") String token,
+      @PathVariable("symbol") String symbol,
+      @PathVariable("range") String range,
+      @PathVariable("date") String date);
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/service/IexClient.java:54]: Nice!

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 8]
```diff
@@ -27,8 +28,8 @@
    *
    * @return a list of all Stock Symbols from IEX.
    */
-  public List<IexSymbol> getAllSymbols() {
-    return iexClient.getAllSymbols();
+  public List<IexSymbol> getAllSymbols(final String token) {
+    return iexClient.getAllSymbols(token);
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/service/IexService.java:33]: This is where I would use the ```mvc.iexToken```

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 9]
```diff
@@ -37,13 +38,27 @@
...
-      return iexClient.getLastTradedPriceForSymbols(symbols.toArray(new String[0]));
+      return iexClient.getLastTradedPriceForSymbols(token, symbols.toArray(new String[0]));
     }
   }
 
+  /**
+   * Get historical prices given stock and time range.
+   *
+   * @param symbol stock symbol for historical prices
+   * @param range time range for historical prices (e.g. 3m, 6m, 5y)
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/service/IexService.java:54]: Can you add documentation about date?

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 10]
```diff
@@ -37,13 +38,27 @@
...
+   * @return a list of historical prices for the symbol/range passed in
+   */
+  public List<IexHistoricalPrice> getHistoricalPrices(final String token, final String symbol,
+      final String range, final String date) {
+    if (date == null) { return iexClient.getHistoricalPrices(token, symbol, range); }
```
Comment (jbrody17) [inline src/main/java/org/galatea/starter/service/IexService.java]: What if symbol is null or and empty string? Range is null or an empty string?

**Review by jbrody17 — COMMENTED** [review]
Overall looks really good. Great job!

File: src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java (hunk 1 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/domain/IexHistoricalPrice.java:18]: It seems that the IEX API historical prices and symbols endpoints return dates as strings: 

image

When I was originally hitting the endpoint, I was getting a lengthy date number (e.g. "updated" key). Is this milliseconds from Jan 1970? 

Let me know if you'd like me to change date back to the Date type or parse the date object to a string instead of receiving the data from the API as a string.

File: src/main/java/org/galatea/starter/domain/IexSymbol.java (hunk 2 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/domain/IexSymbol.java:13]: See above

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 8 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/service/IexService.java:33]: I see... ! :) Thanks for the tip!

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java (hunk 3 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: I removed token as an optional parameter! Instead, IexService passes the token to IexClient.

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java (hunk 4 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/entrypoint/IexRestController.java]: Fixed!

File: src/main/java/org/galatea/starter/entrypoint/IexRestController.java (hunk 5 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/entrypoint/IexRestController.java:57]: Done!

File: src/main/java/org/galatea/starter/service/IexClient.java (hunk 6 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/service/IexClient.java:50]: Done!

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 9 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/service/IexService.java:54]: Done!

File: src/main/java/org/galatea/starter/service/IexService.java (hunk 10 above)
Comment (johncho108) [reply src/main/java/org/galatea/starter/service/IexService.java]: I added and changed the logic in IexService to handle optional range and date parameters. Since the IEX API handles both null and empty strings for range/date parameters identically, I changed this method to pass an empty string to IexClient in both cases. I also added some test cases to demonstrate expected results from optional range and date parameters. 

As for symbol, I think my endpoint is handling null or empty symbol strings appropriately. I added test cases to demonstrate expected results from null and empty symbol strings. 

Given the above changes, tests still pass:

tests_passed

**Review by jbrody17 — APPROVED** [review]
Looks great. Fast turnaround 👍
